add_android_openssl_libraries: Automatically link OpenSSL on argument…

aambrosano opened the pull request:
… targets

aleamb-kdab committed:
add_android_openssl_libraries: Automatically link OpenSSL on argument targets

aambrosano commented:
For the record, `OpenSSL::OpenSSL` is not really the right target. `FindOpenSSL` will actually create `OpenSSL::crypto` and `OpenSSL::ssl`, however linking against those targets is still generating errors at the CMake generation step. Couldn't figure out why yet.

Issam-b commented on android_openssl.cmake:
```diff
@@ -1,26 +1,30 @@
 function(add_android_openssl_libraries)
   if(CMAKE_BUILD_TYPE STREQUAL "Debug")
-    set(ssl_root_path ${CMAKE_CURRENT_FUNCTION_LIST_DIR}/no-asm)
+    set(SSL_ROOT_PATH ${CMAKE_CURRENT_FUNCTION_LIST_DIR}/no-asm)
```
keep local variable as lowercase

Issam-b commented on android_openssl.cmake:
```diff
@@ -1,26 +1,30 @@
+    set(OPENSSL_INCLUDE_DIR ${SSL_ROOT_PATH}/ssl_1.1/include)
   endif()
 
   set_target_properties(${ARGN} PROPERTIES QT_ANDROID_EXTRA_LIBS
```
I think this can also be under the ${ARGN} loop

Issam-b commented on android_openssl.cmake:
```diff
@@ -1,26 +1,30 @@
-    if(NOT OPENSSL_ROOT_DIR)
-      set(OPENSSL_ROOT_DIR ${SSL_ROOT_PATH}/ssl_3/${CMAKE_ANDROID_ARCH_ABI})
-    endif()
     list(APPEND android_extra_libs
```
we no longer need to assign this android_extra_libs list if we're doing it for OPENSSL_SSL_LIBRARY and OPENSSL_INCLUDE_DIR, we can use those latter two variables directly for QT_ANDROID_EXTRA_LIBS

Issam-b commented on android_openssl.cmake:
```diff
@@ -1,26 +1,30 @@
+    set(SSL_ROOT_PATH ${CMAKE_CURRENT_FUNCTION_LIST_DIR})
   endif()
 
   if(Qt6_VERSION VERSION_GREATER_EQUAL 6.5.0)
```
A bit simpler re-write for this could be:
`
if(Qt6_VERSION VERSION_GREATER_EQUAL 6.5.0)
    set(OPENSSL_CRYPTO_LIBRARY ${ssl_root_path}/ssl_3/${CMAKE_ANDROID_ARCH_ABI}/libcrypto_3.so)
    set(OPENSSL_SSL_LIBRARY ${ssl_root_path}/ssl_3/${CMAKE_ANDROID_ARCH_ABI}/libssl_3.so)
    set(OPENSSL_INCLUDE_DIR ${ssl_root_path}/ssl_3/include)
  else()
    set(OPENSSL_CRYPTO_LIBRARY ${ssl_root_path}/ssl_1.1/${CMAKE_ANDROID_ARCH_ABI}/libcrypto_1_1.so)
    set(OPENSSL_SSL_LIBRARY ${ssl_root_path}/ssl_1.1/${CMAKE_ANDROID_ARCH_ABI}/libssl_1_1.so)
    set(OPENSSL_INCLUDE_DIR ${ssl_root_path}/ssl_1.1/include)
  endif()

  find_package(OpenSSL REQUIRED)

  foreach(TARGET ${ARGN})
    if (TARGET ${TARGET})
        set_target_properties(${TARGET} PROPERTIES APPEND
          QT_ANDROID_EXTRA_LIBS ${OPENSSL_CRYPTO_LIBRARY} ${OPENSSL_SSL_LIBRARY})
        target_link_libraries(${TARGET} PUBLIC OpenSSL::OpenSSL)
    endif()
  endforeach()
`

Issam-b replied:
actually the set_target_properties() line can be changed to
`
set_property(TARGET ${TARGET} APPEND PROPERTY QT_ANDROID_EXTRA_LIBS
            ${OPENSSL_CRYPTO_LIBRARY} ${OPENSSL_SSL_LIBRARY})
`

Issam-b reviewed (changes requested):
> For the record, `OpenSSL::OpenSSL` is not really the right target. `FindOpenSSL` will actually create `OpenSSL::crypto` and `OpenSSL::ssl`, however linking against those targets is still generating errors at the CMake generation step. Couldn't figure out why yet.

This doesn't work for me, I get an error:

> CMake Error at ~/Library/Android/sdk/android_openssl/android_openssl.cmake:24 (target_link_libraries):
>  Target "tst_openssl" links to:
> 
>     OpenSSL::OpenSSL
> 
>   but the target was not found.  Possible reasons include:
> 
>     * There is a typo in the target name.
>     * A find_package call is missing for an IMPORTED target.
>     * An ALIAS target is missing.
> 
> Call Stack (most recent call first):
>   CMakeLists.txt:17 (add_android_openssl_libraries)

Issam-b commented:
@aambrosano btw the commit message doesn't explain why such patch is needed, is it trying to fix some bug?

Issam-b commented:
@aambrosano on this

> For the record, OpenSSL::OpenSSL is not really the right target. FindOpenSSL will actually create OpenSSL::crypto and OpenSSL::ssl, however linking against those targets is still generating errors at the CMake generation step. Couldn't figure out why yet.

It doesn't work because the targets should be OpenSSL::Crypto and OpenSSL::SSL not OpenSSL::crypto and OpenSSL::ssl.

aambrosano commented:
@Issam-b it just made sense that we linked against OpenSSL in the targets we passed to `add_android_openssl_libraries`, which we weren't doing. Otherwise we'd need to look up for those libraries twice and that would be redundant. All other changes were just small adjustments.

> @aambrosano btw the commit message doesn't explain why such patch is needed, is it trying to fix some bug?

aambrosano commented:
> @aambrosano on this
> 
> > For the record, OpenSSL::OpenSSL is not really the right target. FindOpenSSL will actually create OpenSSL::crypto and OpenSSL::ssl, however linking against those targets is still generating errors at the CMake generation step. Couldn't figure out why yet.
> 
> It doesn't work because the targets should be OpenSSL::Crypto and OpenSSL::SSL not OpenSSL::crypto and OpenSSL::ssl.

I was just typing targets names off the top of my head when commenting, but was using the right names when coding. That will not work anyway because for some reason targets found within a function are not available outside of the scope of the function. Trying to link against `OpenSSL::SSL` and `OpenSSL::Crypto` will result in an error message like this one:

```
CMake Error in CMakeLists.txt:
  The dependency target "OpenSSL::SSL" of target
  "mytarget_autogen_timestamp_deps" does not exist.
  ```

Even when converting the function to a macro (which has its own caveats anyway) the listing will produce the same error. Will have to think about this one a bit more.

Issam-b commented:
@aambrosano do you still have the issue above? things work right on my side with this PR, I'd like to check that before submitting this.

alcroito commented on android_openssl.cmake:
```diff
@@ -6,21 +6,26 @@ function(add_android_openssl_libraries)
 
-  set_target_properties(${ARGN} PROPERTIES QT_ANDROID_EXTRA_LIBS
-                                           "${android_extra_libs}")
+  find_package(OpenSSL REQUIRED GLOBAL)
```
2 comments here

1) `GLOBAL` option was introduced in cmake 3.24. Any project using a lower cmake version would not be able to use this function anymore. This includes also latest Qt projects, which require a minimum of 3.22.

2) Even if the cmake verison was not an issue, creating global targets is generally not great, because it prevents being able to create local (directory-scoped) targets with the same name, but in different directories.

Issam-b replied:
I took the liberty to remove that, in the last commit, @alcroito does the patch look good for you?

alcroito replied:
Personally, i don't think it looks good, but i'm not the maintainer.

aambrosano mentioned that without `GLOBAL`, the targets are not available in certain cases. I believe him. So for some use cases it likely won't work. Probably has to do with the scope where the function is called, and where the app is created.

Furthermore, doing `find_package` inside functions is frowned upon, because all the variables that `find_package` sets won't be available outside.

Also, I find it strange to do find_package(OpenSSL) to create the targets with pre-filled input from the computed variables.

At that point you could just create the imported targets manually, and avoid the `find_package` call altogether.

aambrosano replied:
Just got to test the patch again, as it broke builds on the project I'm using it on.

The patch doesn't look like it works as intended, because `OpenSSL::*` targets won't be visible from outside the function and therefore the build will fail. Without making them `GLOBAL`, they might as well not exist in the first place as they don't contribute to the final build in any significant way.

I agree it is probably not the best to call `find_package` inside a function. At the same time I find strange to create targets manually when `find_package` could do it for us just fine. But I don't have a strong opinion about it, and since `GLOBAL` is not supported for `find_package` in older CMake versions we might as well create the target ourselves.

A few points I have though, and the main reason why I submitted this PR, are:
- This repo already knows how to set up a reasonable targets for OpenSSL on Android
- Manually creating targets downstream where linking Android projects against OpenSSL is error-prone and repetitive
- In principle, at the previous state of things (and at the current too since the linking part doesn't really work), the user could link against one OpenSSL library (manually crafting targets) and add some different lib as `QT_ANDROID_EXTRA_LIBS` through `add_android_openssl_libraries`. The PR was an attempt to do it in a single step and preventing bugs on user side and unnecessary code duplication across codebases.

I am not sure yet what the best way of doing this is, but I do think it should happen in one place, could be within the same CMake function or just within this repo somehow.

Back to the merged change. What has been merged doesn't solve the problem described above, so I can only see the following ways forward:
- Revert the patch altogether -> WONTFIX
- Keep it as is now, if you like it more (e.g. from a code style POV), and possibly drop the extra targets that have no effect  -> WONTFIX
- Make the targets global -> this is actually a fix, remains to be seen how clean/acceptable that is

If you have other ideas, I am open to suggestions.

Issam-b committed:
Remove GLOBAL from find_package() call

And minor formatting.

Issam-b committed:
Use add_library() to add OpenSSL libraries to the project

If the OpenSSL targets are not defined, add them with add_library() and set OPENSSL_LIB_DIR and OPENSSL_INCLUDE_DIR and store the paths to the shared libraries in the cache so that add_android_openssl_libraries() can re-use them later if OpenSSL targets are already defined.

Issam-b commented:
I pushed a new version that avoids the issues pointed out by @alcroito. This is working correctly for me, @aambrosano does it work in the cases that were problematic for you before?
